lib/framework/procbot.ts

```diff
@@ -55,43 +55,75 @@ export class ProcBot {
 	 * Query properties of the environment and check that they all match.
 	 * Must be properties that DO NOT execute, e.g. process.pid but not process.exit().
 	 * @param filter  List of property/value pairs for process and os that all must match.
-	 * @returns       Whether all properties matched.
+	 * @returns       Promise that resolves to whether all properties matched.
 	 */
-	public static matchEnvironment(filter: ProcBotEnvironmentProperties): boolean {
+	public static matchEnvironment(filter: ProcBotEnvironmentProperties): Promise<boolean> {
 		const matchProcess = _.every(filter.process, (term) => {
 			return JSON.stringify(process[term.property]) === term.value;
 		});
 		const matchSystem = _.every(filter.system, (term) => {
 			return JSON.stringify(os[term.property]) === term.value;
 		});
-		return matchProcess && matchSystem;
+		if (!matchProcess || !matchSystem) {
+			return Promise.resolve(false);
+		} else if (_.isEmpty(filter.package)) {
+			return Promise.resolve(true);
+		}
+		return fsReadFile('./package.json')
```

Comment:
If for some reason the bot isn't run from the root of `resin-procbots`, this has the potential to throw, and the entire framework stops (if not called from within a Promise chain with a catch block).

You probably want to use `require.main` to get the entry point and infer the location of the `package.json` from there. AFAIK we don't run via any sort of package manager atm, so this should be safe.

You also need to add a catch which logs if for some reason the file isn't present or parseable or is run from outside a Promise chain.

Reply:
@hedss re: root.  As far as I can tell `npm start` puts the cwd to the location of package.json

Reply:
@sqweelygig Yes, it does. But process managers (such as `pm2`) abstract the CWD, and there's absolutely no telling how we decide to start the process in the future, esp. moving forwards with K8. :)